text/0000-remove-shell-processes.md

```diff
@@ -0,0 +1,401 @@
+* If the buildpack authors believed that `--urls` should be overridable they could set move the last two arguments from `command` to `args`.
+
+### Example 2 - A Script Process
+When a buildpack upgrade to the new buildpack API, it must convert any existing shell processes into direct processes.
+
+The [Paketo Yarn Start Buildpack](https://github.com/paketo-buildpacks/yarn-start) currently may generate a script processes similar to the following:
+
+```
+[[processes]]
+type = "web"
+command = "pre-start.sh && nodejs server.js && post-start.sh"
+direct = false
+default = false
+```
+
+Using the new API this process look like:
+```
+[[processes]]
+type = "web"
+command = ["bash", "-c", "pre-start.sh && nodejs server.js && post-start.sh"]
+default = false
+```
+
+## User Provided Processes
+
+Currently if the user can specify a custom process dynamically at runtime by setting the container entrypoint to `launcher` directly rather than using a symlink to the launcher, the providing a custom `cmd`. This custom command is executed directly if `cmd` is an array and the first element is `--`. Otherwise the custom command is assumed to be a shell process. In the interest of removing complexity we should do away with the special `--` argument and execute all custom commands directly. 
+
+### Example 1 - A Direct process
+The follow direct commands:
+```
+docker run --entrypoint launcher <image> -- env
+docker run --entrypoint launcher <image> -- echo hello '$WORLD' 
+```
+will become the following, using the new platform API
+```
+docker run --entrypoint launcher <image> env
+docker run --entrypoint launcher <image> echo hello '${WORLD}'
+```
+Previously, in the second command in this example `${WORLD}` would not have been interpolated because this is a direct process; instead the output would include the literal string `$WORLD`. With the changes proposed, `${WORLD}` will now be evaluated, even though the process is direct.
+
+### Example 2 - A Shell Process
+The follow custom shell command:
+```
+docker run --entrypoint launcher <image> echo hello '${WORLD}'
+docker run --entrypoint launcher <image> echo hello '${WORLD:-world}'
+```
+will become the following, using the new platform API
+```
+docker run --entrypoint launcher <image> echo hello '${WORLD}'
+docker run --entrypoint launcher <image> bash -c 'echo hello "${WORLD:-world}"'
+```
+The first command in this example did not need to change to behave as expected with the new API. Previously it was necessary to use a shell process in order to evaluate `${WORLD}`. Now, the shell is unescessary. However, if the user wants to use more features of Bash interpolation they will need to explicitly invoke a shell, as shown in the second command in this example.
+
+### Example 3 - A Script Process
+The follow custom script command:
+```
+docker run --entrypoint launcher <image> 'for opt in $JAVA_OPTS; do echo $opt; done'
```

Comment:
One of the things I worry about in this example is how it might unexpectedly fail. I expect that most users will do the following:

1. upgrade
2. try the exact same command they used before (i.e. no `bash -c`)
3. watch it fail
4. realize they need to add `bash -c`

In step 3, is it possible that something very bad happens if their process kinda works but not as they expected?

Reply:
I agree that they might upgrade and fail but I don't think there is any danger of the process even kinda working. In this case the launcher would be trying to directly invoke an executable called `'for opt in $JAVA_OPTS; do echo $opt; done'`